packages/openapi-v3/src/__tests__/unit/consolidate-schema.unit.ts

```diff
@@ -0,0 +1,195 @@
+// Node module: @loopback/openapi-v3
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {ConsolidateSchemaObjects, OpenAPIObject} from '../..';
+
+describe('ConsolidateSchemaObjects', () => {
+  it('moves schema with title to component.schemas, replace with reference', () => {
+    const inputSpec: OpenAPIObject = {
+      openapi: '',
+      info: {
+        title: '',
+        version: '',
+      },
+      paths: {
+        schema: {
+          title: 'loopback.example',
+          properties: {
+            test: {
+              type: 'string',
+            },
+          },
+        },
+      },
+    };
+    const expectedSpec: OpenAPIObject = {
+      openapi: '',
+      info: {
+        title: '',
+        version: '',
+      },
+      paths: {
+        schema: {
+          $ref: '#/components/schemas/LoopbackExample',
+        },
+      },
+      components: {
+        schemas: {
+          LoopbackExample: {
+            title: 'loopback.example',
+            properties: {
+              test: {
+                type: 'string',
+              },
+            },
+          },
+        },
+      },
+    };
+    expect(ConsolidateSchemaObjects(inputSpec)).to.eql(expectedSpec);
+  });
+
+  it('ignores schema without title property', () => {
+    const inputSpec: OpenAPIObject = {
+      openapi: '',
+      info: {
+        title: '',
+        version: '',
+      },
+      paths: {
+        schema: {
+          properties: {
+            test: {
+              type: 'string',
+            },
+          },
+        },
+      },
+    };
+    expect(ConsolidateSchemaObjects(inputSpec)).to.eql(inputSpec);
+  });
+
+  it('Avoids naming collision', () => {
+    const inputSpec: OpenAPIObject = {
+      openapi: '',
+      info: {
+        title: '',
+        version: '',
+      },
+      paths: {
+        schema: {
+          title: 'loopback.example',
+          properties: {
+            test: {
+              type: 'string',
+            },
+          },
+        },
+      },
+      components: {
+        schemas: {
+          LoopbackExample: {
+            title: 'Different LoopbackExample exists',
+            properties: {
+              testDiff: {
+                type: 'string',
+              },
+            },
+          },
+        },
+      },
+    };
+    const expectedSpec: OpenAPIObject = {
+      openapi: '',
+      info: {
+        title: '',
+        version: '',
+      },
+      paths: {
+        schema: {
+          $ref: '#/components/schemas/LoopbackExample2',
+        },
+      },
+      components: {
+        schemas: {
+          LoopbackExample: {
+            title: 'Different LoopbackExample exists',
+            properties: {
+              testDiff: {
+                type: 'string',
+              },
+            },
+          },
+          LoopbackExample2: {
+            title: 'loopback.example',
```

Comment:
This a an interesting edge case! I am glad you included it in your pull request.

I am not sure what's the best or the right way how to handle the situation, I guess your proposal is a good start.

Reply:
Thought is would be prudent with new enhancers allowing developers to 'get creative'.  Might need to be more edgy in future!